### ISSUE TITLE
provider/rabbitmq: Arguments Map values are all strings

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-rabbitmq#1]
_This issue was originally opened by @dmportella as hashicorp/terraform#10186. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

### Terraform Version
![Version](https://cloud.githubusercontent.com/assets/676817/20384722/beea35a8-acac-11e6-8cf0-30051c1e7c03.png)
> v0.7.11

### Affected Resource(s)
Please list the resources as a list, for example:
- RABBITMQ_BINDING
- RABBITMQ_EXCHANGE
- RABBITMQ_QUEUE
- RABBITMQ_POLICY (Possibly)

> If this issue appears to affect multiple resources, it may be an issue with Terraform's core, so please mention this.

It is possible that its `core` because the Map Types `map[string]interface{}` see and return the values as strings when in the resource they have been expressed as integers or other types.

### Terraform Configuration Files
```
# Configure the Docker provider
provider "docker" {
	host = "unix:///var/run/docker.sock"
}

resource "docker_image" "rabbitmq" {
	name = "rabbitmq:3.6.5-management"
}

resource "docker_container" "rabbitmq" {
	image = "${docker_image.rabbitmq.latest}"
	name = "rabbitmq"

	env = ["RABBITMQ_DEFAULT_USER=guest", "RABBITMQ_DEFAULT_PASS=guest"]

        ports {
            internal = 15672
            external = 15672
            ip = "172.17.0.1"
        }

	restart = "always"
}

resource "null_resource" "wait" {
	depends_on = ["docker_container.rabbitmq"]

	provisioner "local-exec" {
		command = "echo 'Sleeping for 5...' && sleep 5"
	}
}


# Configure the RabbitMQ provider
provider "rabbitmq" {
	endpoint = "http://172.17.0.1:15672"
	username = "guest"
	password = "guest"
}

# Create a virtual host
resource "rabbitmq_vhost" "vhost_1" {
	depends_on = ["null_resource.wait"]

	name = "vhost_1"
}

resource "rabbitmq_permissions" "guest" {
	depends_on = ["null_resource.wait"]

	user = "guest"
	vhost = "${rabbitmq_vhost.vhost_1.name}"
	permissions {
		configure = ".*"
		write = ".*"
		read = ".*"
	}
}

resource "rabbitmq_exchange" "riverstyx" {
	depends_on = ["null_resource.wait"]
	
	name = "riverstyx"
	vhost = "${rabbitmq_permissions.guest.vhost}"
	settings {
		type = "fanout"
		durable = false
		auto_delete = true
	}
}

resource "rabbitmq_queue" "debug" {
	depends_on = ["null_resource.wait"]
	
	name = "debug-queue"
	vhost = "${rabbitmq_permissions.guest.vhost}"
	settings {
		durable = false
		auto_delete = true
		arguments = {
			x-message-ttl = 60000
		}
	}
}

resource "rabbitmq_binding" "binding" {
	depends_on = ["null_resource.wait"]
	
	source = "${rabbitmq_exchange.riverstyx.name}"
	vhost = "${rabbitmq_permissions.guest.vhost}"
	destination = "${rabbitmq_queue.debug.name}"
	destination_type = "queue"
	routing_key = "*"
	properties_key = "%23"
}
```

### Debug Output
Gist for Plan, Apply and debug apply: https://gist.github.com/dmportella/564505d677e5147f871771959fdd9c7a

### Panic Output
There was no panic log.

### Expected Behavior
map[string]interface{} values as they are marshalled into JSON should honor integer values and booleans.

For example:

```
# HTC
arguments = {
	x-message-ttl = 60000
}
```

Should out put a call to the Rabbitmq API with the body in json as:
```
#JSON
{
      "x-message-ttl": 60000
}
```
### Actual Behavior
What actually happened?
Json output and sent to rabbitmq API is all strings when rabbit api expects integers so it throws a `Bad Request 400`.

```
#JSON
{
      "x-message-ttl": "60000"
}
```

### Steps to Reproduce
Using the TF file pasted above you should be able to get the error by just doing an apply.
The terraform plan is completely self contained and it uses docker to create the rabbit instance.

1. `terraform apply`

### Important Factoids
Rabbitmq API arguments like x-message-ttl all are reserved keywords in rabbit that checks for the field type like if it is not integer it will fail the request with a `Bad Request 400`.

**Comment by dmportella** [comment]
i was wondering what happen new home for providers

**Comment by cpuspellcaster** [comment]
I know there was some discussion in the [previous issue](https://github.com/hashicorp/terraform/issues/10186) about creating a raw interface for passing in argument JSON directly, but what about explicitly re-typing x-arguments to ints if the API is known to only accept ints?

I'm thinking that logic could be added here in this block: https://github.com/terraform-providers/terraform-provider-rabbitmq/blob/master/rabbitmq/resource_queue.go#L164 doing something like [this](https://play.golang.org/p/zUejaCwcXF). The x-arguments that I'm aware of that take ints are:
- x-message-ttl
- x-expires
- x-max-length
- x-max-length-bytes
- x-max-priority

This is becoming a blocker for something that I'm working on, so I have no problem diving in here.

**Comment by dmportella** [comment]
The Only thing is that the arguments are open to the user to put what ever they want if someone has a custom argument that is a int, float etc moving to terraform rabbit provider is going to be a problem for them because they would be forced to change their data type.

**Comment by kendavis2** [comment]
This limitation preventing my project from using this awesome provider as well.

Per the documentation... "You can declare priority queues using the  x-max-priority argument. This argument should be an `integer` indicating the maximum priority the queue should support." [docs](https://www.rabbitmq.com/priority.html)

I am new to Terraform, but could a type option be included for each argument defaulting to a string if not supplied? I know the RabbitMQ Management Interface allows the user to choose a type.

```
resource "rabbitmq_queue" "queue_name" {
  name  = "queue_name"
  vhost = "vhost_name"

  settings {
    durable     = false
    auto_delete = false
    arguments = {
      x-max-priority = {
         value = 9
         type = "string"
      }   
    }
  }
}
```

**Comment by rhoml** [comment]
Hello Folks,

Is there any ETA on this regards, I was wondering if this should be done at the Terraform level or at the provider level. This not only affects queues, but also policies

**Comment by cpuspellcaster** [comment]
@rhoml I believe this should be handled at the provider level.

My regrets: I indicated that I would work on this, but I haven't had time to follow up. I did complete an initial fix for my own needs, check it out here: https://github.com/onoffleftright/terraform-provider-rabbitmq/commit/f8a79fa5308c8ced4c425b512f7905d5c63167a9

I haven't opened a PR because I didn't think it was a great solution to the problem. @kendavis2 points out that basically any argument can be a string/number/boolean type. My approach just looks for the argument names I mentioned above and type asserts them to ints.

I'm thinking about a way to declare a type in the arguments block without breaking the current interface. Changing each argument from a key/value to a structured block is going to break pre-existing stuff out there that folks have written, I think.

I haven't myself seen any other patterns in the Terraform community that tackles the type issue we have here. I'd like to propose this and get a bit more consensus from the community:

- Let's keep the same key/value pattern in the argument blocks, and default to typing values are strings.
- If the value needs to be passed to RabbitMQ's API as a different type than string, let's define that in the value using some sort of delimiter.

e.g.

```
arguments = {
    custom-arg = "this is a default, so it's a string"
    x-something-ttl = "60000:number"
    convert-something = "true:boolean"
}
```

I don't think that the `:` character is a great delimiter in this case, there is probably a better pattern. I just wanted to illustrate what I'm thinking here.

**Comment by dmportella** [comment]
Yeah honestly is annoying.

**Comment by jtopjian** [comment]
Hi all,

I've just opened #6 which proposes a solution to this. The solution implements the suggestion made [here](https://github.com/hashicorp/terraform/issues/10186#issuecomment-306592649) to define the arguments as a JSON string.

A new parameter called `json_arguments` is used so the existing `arguments` is retained for backwards compatibility.

Please let me know if anyone has any thoughts or comments about this. As well, if you're able to test this, that would be awesome. :)

**Comment by dmportella** [comment]
if its all done it can be closed

**Comment by jtopjian** [comment]
Fixed by #6

**Comment by jtopjian** [comment]
Hi all,

Thank you for your patience with this issue. The fix has been merged. Additionally, a new release of this provider has just been created, which includes this fix.

Do let us know if you run into any further issues.

**Comment by kishorenc** [comment]
@jtopjian Is the merged fix meant for the queue resource only or does it apply to other resources like `rabbitmq_policy`? For e.g. I am facing the same error for the following `rabbitmq_policy`:

```
resource "rabbitmq_policy" "limit-pollingnode" {
  provider = "rabbitmq.rabbitnode"
  name     = "limit queue"
  vhost    = "main"

  policy {
    pattern  = "^queuename"
    priority = 0
    apply_to = "queues"

    definition {
      max-length = 10000
    }
  }
}
```

`max-length` must be an integer, but is probably sent as a string, leading to a `400` error.

**Comment by jtopjian** [comment]
@kishorenc The fix in #6 was only meant for the `rabbitmq_queue` resource. Would you be able to open a new issue to describe what you're seeing?

**Comment by kishorenc** [comment]
@jtopjian I have create a new issue here: https://github.com/terraform-providers/terraform-provider-rabbitmq/issues/12
